lxd/container_lxc.go

```diff
@@ -6422,10 +6425,11 @@ func (c *containerLXC) insertProxyDevice(devName string, m types.Device) error {
 		return fmt.Errorf("Error occurred when starting proxy device: %s", err)
 	}
 
+	go cmd.Wait()
```

Comment:
What does that do? Were you getting zombies if you didn't have a go routine waiting for the process to exit?

Reply:
Yeah I was getting defunct zombie processes without that wait.

Reply:
Ok, so the problem here other than the wasted goroutine is that it means that the proxy isn't backgrounded and so if LXD was to crash, it'd take those proxy processes with it.

I think what we should be doing here is making the proxy process background itself.
That means adding a new argument to it with the location of where to write its pid file and have it (probably in the C loader) double-fork, close stdin, stdout, stderr, ... as a normal daemon would.
I'd also add a "-f" option to the proxy to have it run in the foreground for debugging.

Reply:
That'd effectively make the "lxd forkproxy" processes behave like the "dnsmasq" processes that we spawn, allowing them to keep running even if the main lxd process dies (or is restarted).

Reply:
Yes, we were getting zombies when we don't wait for the command. Although all the exited proxy processes are reaped eventually when lxd exits, we just thought this is cleaner. We could potentially move this to killProxyProc() after we send a SIGTERM signal. What would you suggest?

Reply:
What exactly did you mean by the "double-fork"? Also, would that mean that the proxy would write its own PID file in the CGO code and not in the Go code?

Reply:
Hmm, since the parent is going away double forking shouldn't be necessary.

So yeah, you'd want the cgo code to do a fork(), exit() of the parent and then setsid() of the child in the case where it's about to become a daemon (and not in the case where it's just about to send an fd back and exit).

You may want to look at doing this in Go, if it's something other projects do, then doing it in Go should make it more readable, otherwise, it's a pretty trivial amount of C code to do it.

As for the pid file, it doesn't really matter where you write it, so I suspect doing it from the Go code will be easier.

Reply:
Would it make more sense for the proxy to be changed fork and exit or to just call `os.FindProcess()` and then call `process.Wait()` to reap the zombie child in `killProxyProc()`?

Reply:
fork and exit is (daemonizing) is the only way you're going to avoid the proxy dying when LXD dies.

Reply:
Would we actually need to fork the proxy during the first time the CGO code runs and then call `setsid()` in the child process? Could we not just call `setsid()` without forking, which would separate the proxy from LXD's and *should* give the correct behavior?

If we go with the forking approach, then we technically would not need `SpawnCommand()` since we can just wait for the parent to `exit()` to return control to LXD while the child proxy keeps running in the background. We could have the PID of the child process be returned to LXD via the exit status of the parent proxy or capture it from STDOUT in `RunCommand()`.

If we do not need to fork, then we would still need `SpawnCommand()` to background the proxy which would call `setsid()` and separate itself from LXD.

Both approaches would need to do different things to allow the proxy process to be foregrounded also. For the forking approach that does not need `SpawnCommand()`, the proxy would just not fork the child and do the proxying itself. For the non-forking approach that uses `SpawnCommand()`, it would need to call `cmd.Wait()` on the return value of `SpawnCommand()`.

If the client chooses to foreground the proxy process, would all of the `logger.Debugf()` calls need be `fmt.Printf()` calls? As in, one of the args passed into the proxy would be a flag that determines where to send the output?

Reply:
setsid() would fix the child process dying on lxd exit issue but it'd still require you to have the background .Wait() goroutine as lxd would still be your parent and so would need to handle SIGCLD to avoid the process becoming a zombie.

I think forking is the best bet here and indeed, you could then drop SpawnCommand and just use RunCommand as we do everywhere else.

You could print the pid through stdout and grab it that way, but it may also just be simpler to have the proxy write its own pid into the pidfile directly, that'd be one less thing to wrap around RunCommand in the lxd code.

I'd say not to worry about the foregrounded case right now, we can add that later if we find it useful to debug a particular issue. It wouldn't be an option that LXD itself uses.
If we do add that option, then logging should just work as I believe we setup the logging handler in main.go for every subcommands.

Reply:
While I was trying to daemonize the proxy process, I ran into an issue that seems to have occurred before in the LXD codebase. `RunCommand()` hangs because it is executing a process that forks a daemon. Here is the [Golang issue](https://github.com/golang/go/issues/13155) to which I am referring. You are probably familiar with it since you merged a [PR](https://github.com/lxc/lxd/pull/1488/files)  that involved the issue. 

So there is obviously a workaround to this issue. We can close STDIN, STDOUT, and STDERR in the CGO code or at the beginning of the Go code in the proxy. The problem with it is that closing all those descriptors means that there will be no output returned from `RunCommand()` when executing the proxy. There will be nothing letting the user know if the daemon failed to spawn from `fork()`.  Obviously, the Go code has access to `logger.Debugf()` and `logger.Errorf()` for its logging but the C code does not. 

Also, using this workaround, the proxy is forced to write its own PID file since there is no way to return the PID to `RunCommand()`. There is also error handling/reporting that would need to be done when writing the PID file. If we passed the file path to the proxy, in the Go code, we could create the PID file and be able to log any errors that occurred.

Essentially, the Go code is fine since we can change all of the `fmt.Errorf()` calls to `logger.Errorf()` calls and leave the `logger.Debugf()` calls. The CGO code will not have any error reporting whatsoever in the parent or the child daemon since the file descriptors were closed.

Is there anything that I am missing here that would get this to work?

Also, regarding the non-forking approach where we just call `setsid()` immediately, we would not necessarily need to have the `go cmd.Wait()`. When we are killing the proxy, we have the PID of the process. We can call `os.FindProcess(proxyPid).Wait()` which should reap the proxy. That way there would not be a zombie process hanging around and we would not need to have wasted goroutines running. This approach does not daemonize the proxy but would have much better error reporting than the above-mentioned workaround for daemonizing.

Reply:
What we've been doing for the other fork* subcommands is have them write to a logfile. In this case, this would be a new forkproxy.log that we could have the proxy set its stdout and stderr to.

Reply:
Cool. I will look at the other subcommands and do that. Thanks!

Reply:
Would there be a different logfile for each proxy device? If so, should there be a separate folder under the log path for devices? Something like /var/log/lxd/<container>/devices/proxy.<name>.log. And just to be clear, it would pipe all output to the log file so all the `logger.Debugf()` calls could be replaced with `fmt.Printf()` calls that are redirected to the file.